File: nfc_proof_of_concept/packages/flutter_hce/lib/app_layer/file_access/serializers/capability_container_serializer.dart
import '../../field.dart';
import '../fields/capability_container_fields.dart';
import 'tlv_block_serializer.dart';

class CapabilityContainer extends ApduSerializer {
  late final CcLenField cclen;
  final CcMappingVersionField version = CcMappingVersionField.v2_0;
  final CcMaxApduDataSizeField mLe;
  final CcMaxApduDataSizeField mLc;
  final List<FileControlTlv> fileDescriptors;

  CapabilityContainer({
    required this.fileDescriptors,
    int? maxResponseSize,
    int? maxCommandSize,
  })  : mLe = CcMaxApduDataSizeField.mLe(maxResponseSize ?? 0x00FF),
        mLc = CcMaxApduDataSizeField.mLc(maxCommandSize ?? 0x00FF),
        super(name: "Capability Container") {
    if (fileDescriptors.isEmpty) {
      throw ArgumentError('CapabilityContainer must have at least one file descriptor.');
    }
  }

  @override
  void setFields() {
    int totalTlvLength = 0;
    for (final descriptor in fileDescriptors) {
      totalTlvLength += descriptor.length;
    }
    
    // Header is 7 bytes: CLEN(2) + Version(1) + MLe(2) + MLc(2)
    final int totalLength = 7 + totalTlvLength;
    cclen = CcLenField(totalLength);

    fields = [
      cclen,
      version,
      mLe,
      mLc,
      ...fileDescriptors,
    ];
  }
}

File: nfc_proof_of_concept/packages/flutter_hce/android/src/main/kotlin/com/viridian/flutter_hce/app_layer/file_access/serializers/CapabilityContainerSerializer.kt
package com.viridian.flutter_hce.app_layer.file_access.serializers

import com.viridian.flutter_hce.app_layer.ApduSerializer
import com.viridian.flutter_hce.app_layer.file_access.fields.*

class CapabilityContainer(
    private val fileDescriptors: List<FileControlTlv>,
    maxResponseSize: Int? = null,
    maxCommandSize: Int? = null
) : ApduSerializer("Capability Container") {
    
    private val cclen: CcLenField
    private val version: CcMappingVersionField = CcMappingVersionField.v2_0
    private val mLe: CcMaxApduDataSizeField = CcMaxApduDataSizeField.mLe(maxResponseSize ?: 0x00FF)
    private val mLc: CcMaxApduDataSizeField = CcMaxApduDataSizeField.mLc(maxCommandSize ?: 0x00FF)

    init {
        require(fileDescriptors.isNotEmpty()) { "CapabilityContainer must have at least one file descriptor." }
        var totalTlvLength = 0
        for (descriptor in fileDescriptors) {
            totalTlvLength += descriptor.length
        }

        // Header is 7 bytes: CCLEN(2) + Version(1) + MLe(2) + MLc(2)
        val totalLength = 7 + totalTlvLength
        cclen = CcLenField(totalLength)

        // Register in order
        register(cclen)
        register(version)
        register(mLe)
        register(mLc)
        for (descriptor in fileDescriptors) {
            register(descriptor)
        }
    }
}
